translate/compiler.go

```diff
@@ -37,19 +44,102 @@ func CompileContract(c ast.Contract) (Bytecode, error) {
 
 	memTracer := NewMemEntryTable()
 
+	funcMap := FuncMap{}
+
 	for _, f := range c.Functions {
+		funcMap.Declare(*f, *bytecode)
+
 		if err := compileFunction(*f, bytecode, memTracer); err != nil {
 			return *bytecode, err
 		}
 	}
 
-	if err := generateFuncJumper(bytecode); err != nil {
+	if err := expectFuncJumper(c, funcMap); err != nil {
+		return *bytecode, err
+	}
+
+	if err := generateFuncJumper(c, bytecode, funcMap); err != nil {
 		return *bytecode, err
 	}
 
 	return *bytecode, nil
 }
 
+// Expects a size of the function jumper and updates the function map.
+func expectFuncJumper(c ast.Contract, funcMap FuncMap) error {
+	expectedJpr := NewBytecode()
+	if err := generateFuncJumper(c, expectedJpr, funcMap); err != nil {
```

Comment:
`generateFuncJumper` here is for calculating function jumper bytes size? 

If so, in my intuition we can calculate the fn jumper size right away(because fn sel is 4 bytes), not after generating function jumper.

I comment this because I'm a little bit consufed with `generateFuncJumper`

Reply:
Yes, you're right. I had concern to them, but i couldn't any solution.

Function Jumper logic is the following.

First, compile functions. This means compiling a logic for each function and find the location of that function in byte code. 
Second, expect function jumper.
...

Suppose that the byte code is `00616263` and the start of function to compile (`func foo`) is `63`. Then, in that byte code, `func foo` location is 3. (0: `00`, 1: `61`, 2: `62`, 3: `63`) So in `FuncMap`, `FuncMap["foo"]` is `3`. But if function jumper is added in front of the byte code, the function location is changed. For example, function jumper is `112233`, the byte code should be `11223300616263` and `FuncMap["foo"]` should be `6`. So, I have to update `FuncMap` after adding the function jumper to byte code. In this time, a problem occur. To update `FuncMap`, i must know a size of function jumper. But, to know the function jumper, I generate a function jumper. Then, you might think that 'Compile functions, generate function jumper, update `FuncMap`, then combine function jumper and byte code!'. But, After updating `FuncMap`, I regenerate a function jumper. Because, as changing `FuncMap`, the locations of functions are also changed. 

So, compile functions, generate function jumper to expect the size of function jumper, update `FuncMap`, generate function jumper with updated `FuncMap`, and combine.

If you have a good idea, please tell me!! :)

Reply:
OK I understand your intention about this function. 

If my understanding is right, this function's job is 
(1) create function jumper (just for calculating its size) then  
(2) relocate function map based on its size

So `relocateFuncMapEntry` might be better I think.

Reply:
And how about receiving function jumper as arguments or just its size?
```go
func expectFuncJumper(funcMap FuncMap, fnJmprSize Size) error { ... }
```

I thought that we don't need to create function jumper, just receive function jumper from outside is enough (and even `ast.Contract` is not needed)

And if we decided `expectFuncJumper` receive function jumper size from arguments, I think we can change the logic in **line 57 ~ 61**. 

```go
if err := expectFuncJumper(c, funcMap); err != nil {
	return *bytecode, err
}

if err := generateFuncJumper(c, bytecode, funcMap); err != nil {
	return *bytecode, err
}
```
to
```go
if err := generateFuncJumper(c, bytecode, funcMap); err != nil {
	return *bytecode, err
}
if err := expectFuncJumper(funcMap, len(bytecode)); err != nil {
	return *bytecode, err
}
```